### PR TITLE
CONTRIBUTING.md: tune up to section formatting + release etc workflow draft

### PR DESCRIPTION
**Pull request by yarikoptic**
If needed - can split into two PRs (or just merge first commit into master directly) -- first commit is kinda obvious.  
2nd one is aiming to establish some formalization of the workflow so we could reach 1.0 (#4585)  and is largely based on current workflow and ideas in (closed) #4238. It might not yet be complete, and not as detailed as #4238 but I think it would be valuable to agree on these basic items first before expanding/formalizing it further.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -535,6 +535,32 @@ Refer datalad/config.py for information on how to add these environment variable
 
 # Release(s) workflow
 
+## Branches
+
+- `master`: changes toward the next `MAJOR.MINOR.0` release (e.g., `0.13` as of time of writing).
```
Comment (kyleam) [inline CONTRIBUTING.md:540]: I think the "(e.g., ...)" bit can safely be dropped (same for the instance below).  If it stays, this example tag should have a matching format by appending a ".0".

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
 - *DATALAD_ALLOW_FAIL*:
   Instructs `@never_fail` decorator to allow to fail, e.g. to ease debugging.
 
-# Changelog section
+# Release(s) workflow
+
+## Branches
+
+- `master`: changes toward the next `MAJOR.MINOR.0` release (e.g., `0.13` as of time of writing).
+  It is also used for release candidates (with `rcX` suffix)
```
Comment (kyleam) [inline CONTRIBUTING.md:541]: Perhaps "Release candidates (tagged with an `rcX` suffix) are cut from this branch".

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+- `maint`: bug fixes and minor enhancements for the latest released `MAJOR.MINOR.PATCH`
```
Comment (kyleam) [inline CONTRIBUTING.md:542]: I'd prefer to drop the "minor enhancements" part.  Of course, they can be argued on a case by case base, but I don't see any reason to suggest they should go to maint.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+ (e.g., `0.12.7` as of time of writing)
+- `MAJOR.MINOR`: generally not used, unless some bugfix release with a crticial bugfix is needed.
```
Comment (kyleam) [inline CONTRIBUTING.md:544]: If we were to need to do this, I'd vote for a "maint-" in front.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+
+## Workflow
+
+- upon release of `MAJOR.MINOR.0`, `maint` branch need to be fast-forwarded to that release
+- bug fixes to functionality released within the `maint` branch, should (HIGHLY RECOMMENDED) be
+  submitted against `maint ` branch
+- cherry-picking fixes from `master` into `maint` is allowed where needed
+- `master` branch accepts PRs with new functionality
+- `master` branch merges `maint` as frequently as possible/needed
```
Comment (kyleam) [inline CONTRIBUTING.md:553]: "as possible/needed" seems to be presenting two conflicting options.  FWIW my handling and preference for it is more towards the "needed".

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+- `master`: changes toward the next `MAJOR.MINOR.0` release (e.g., `0.13` as of time of writing).
+  It is also used for release candidates (with `rcX` suffix)
+- `maint`: bug fixes and minor enhancements for the latest released `MAJOR.MINOR.PATCH`
+ (e.g., `0.12.7` as of time of writing)
+- `MAJOR.MINOR`: generally not used, unless some bugfix release with a crticial bugfix is needed.
+
+## Workflow
+
+- upon release of `MAJOR.MINOR.0`, `maint` branch need to be fast-forwarded to that release
+- bug fixes to functionality released within the `maint` branch, should (HIGHLY RECOMMENDED) be
```
Comment (kyleam) [inline CONTRIBUTING.md:549]: "should" alone seems sufficient.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+  submitted against `maint ` branch
+- cherry-picking fixes from `master` into `maint` is allowed where needed
+- `master` branch accepts PRs with new functionality
+- `master` branch merges `maint` as frequently as possible/needed 
+
+## Code flow
```
Comment (kyleam) [inline CONTRIBUTING.md:555]: This seems like it should certainly be in a separate commit, but probably a separate PR, since I don't recall seeing a proposal that matches this.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+
+## Branches
+
+- `master`: changes toward the next `MAJOR.MINOR.0` release (e.g., `0.13` as of time of writing).
+  It is also used for release candidates (with `rcX` suffix)
+- `maint`: bug fixes and minor enhancements for the latest released `MAJOR.MINOR.PATCH`
+ (e.g., `0.12.7` as of time of writing)
+- `MAJOR.MINOR`: generally not used, unless some bugfix release with a crticial bugfix is needed.
+
+## Workflow
```
Comment (kyleam) [inline CONTRIBUTING.md:546]: I think it's fine to have this section, but I think more importantly we should resurrect and tweak the text we had in .github/PULL_REQUEST_TEMPLATE.md before 82e9d6a57 (MNT: Drop "Choose a base" item from PR template checklist, 2020-01-16).

File: CONTRIBUTING.md (hunk 3 above)
Comment (kyleam) [inline CONTRIBUTING.md:542]: You're inconsistent across these entries with your spelling of "bug fix".

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -543,10 +533,47 @@ Refer datalad/config.py for information on how to add these environment variable
...
+- `master` branch merges `maint` as frequently as possible/needed 
+
+## Code flow
+
+- new functionality is initially introduced outside of the `datalad.core.` namespace
+- new functionality should mature for at least a single full `MAJOR.MINOR` release until considered
+  for migration to `datalad.core.`
+  - convenience shim imports remain in-place in old locations for at least a single `MAJOR.MINOR` cycle
+  - they should be accompanied with deprecation warnings to trigger downstream code fixups
+  - our internal code should get adjusted immediately upon migration and testing should fail if was not
```
Comment (kyleam) [inline CONTRIBUTING.md:562]: AFAIK we don't currently have a build that fails if there are internal deprecation warnings.  Within the context of a CONTRIBUTING, I think "testing should fail" will be unclear (is it describing something the contributor needs to actively do? something they should expect will happen with the existing setup? ...).  And writing that makes me think that maybe CONTRIBUTING isn't the best place for this section.

Stepping back, I think we might want to reshape CONTRIBUTING with the goal of making it a bit more useful to first-time contributors, as they are likely the ones that sit down and read it.  I think it'd make sense to find another home for the more detailed reference material that is currently in CONTRIBUTING.

File: CONTRIBUTING.md (hunk 4 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:544]: On one hand - sounds good. On the other 
- we already have such branches for past releases (well - they could be renamed I guess to follow)
- I thought to keep possibility open for the bright future where we might have a more thorough road map, and decide to accept/review prs for future releases, like already happening, without accumulating them unmerged until release. (For that though we would've need to abandon master/maint for clarity). But hey - if such dire need arises, we indeed could rename branches again.

So, I guess I am ok either way. But it reminded me that if we follow current idea - I should add an explicit statement that changes toward release one after the one to come should aim master but will not be merged until release comes, and should be assigned to the corresponding next milestone

File: CONTRIBUTING.md (hunk 7 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:555]: Ok. I could strip it off and postpone (if it feels needing a separate heavy discussion) a pr with it until we settle on previous items.

File: CONTRIBUTING.md (hunk 9 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:562]: Yes, I felt the same somewhat and thought to place it somewhere in docs. But then I also hate searching through docs for the right document and thought that the trail if this document would be appropriate. Also it is somewhat difficult to make a clear separation of what is useful for new vs old contributors. But indeed release management is probably clearly the one which could be elsewhere

File: CONTRIBUTING.md (hunk 4 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:544]: Oi, fresh related topic: https://www.google.com/amp/s/www.zdnet.com/google-amp/article/github-to-replace-master-with-alternative-term-to-avoid-slavery-references/

**Comment by yarikoptic** [comment]
Thank you @kyleam for the review! I will incorporate suggested obvious fixes in short time.
I would have really appreciated if you had used suggestion mode for them (where possible) - that would've allowed for immediate review and acceptance even while on the phone, and having them marked resolved automagically.

File: CONTRIBUTING.md (hunk 4 above)
Comment (kyleam) [reply CONTRIBUTING.md:544]: > * we already have such branches for past releases (well - they could be renamed I guess to follow)

I think ideally the old, inactive branches we have (0.5.x, 0.9.x, 0.11.x) should just be deleted.  If we decide for some reason that we need to make a release off of them, we can branch off of the tag.

> * I thought to keep possibility open for the bright future where we might have a more thorough road map, and decide to accept/review prs for future releases, like already happening, without accumulating them unmerged until release.

If I understand your suggestion, it's something I'd be opposed to, but that's a separate proposal and discussion.

However, re "like already happening": We (in particular me) botched this rc.  I didn't realize you were opposed to releasing `push` in its current form.  We shouldn't have entered the rc phase when things were not settled internally.  It makes what should be a short period drag on indefinitely.

Comment (yarikoptic) [reply CONTRIBUTING.md:544]: > However, re "like already happening": 

I had in mind PRs like #4597 which are to be  "scheduled" for the next release

<details>
<summary>I think the `push` part is not really related to this PR thus collapsed here</summary>

> We (in particular me) botched this rc. I didn't realize you were opposed to releasing push in its current form.

Hm... you can prove me wrong, but I thought that I never actively "opposed" - that is why the only issue in 0.13.0 milestone ATM is https://github.com/datalad/datalad/pull/4453 but I will remove it from any milestone since I do not think we have reached or will reach any agreement on that any time soon and it is not really a critical issue.  I did express my dislikes though, and stated that we would likely need to break API if we release `push` in current form.  So I think it was just a common sense and delaying the release while we hadn't agreed on the API.

</details>

Comment (kyleam) [reply CONTRIBUTING.md:544]: > I had in mind PRs like #4597 which are to be "scheduled" for the next release

Yes.  My point is that it's being held up by an rc that should not have taken as long as it has.  And the reason it is taking so long is we entered the rc period without realizing there disagreements about `push` in the form it was in at rc1.

**Comment by yarikoptic** [comment]
Dear @kyleam , could you please tune it up to your liking/comments?  I have ran out of time :-(

File: CONTRIBUTING.md (hunk 1 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:540]: was dropped

File: CONTRIBUTING.md (hunk 4 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:544]: I would still vote for no `maint-` prefix for the reasons stated above. So we need one more vote from other @datalad/developers to break the tie.

File: CONTRIBUTING.md (hunk 8 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:546]: This is for `Release Management` section, so not sure how PR template relates to that.

File: CONTRIBUTING.md (hunk 4 above)
Comment (yarikoptic) [reply CONTRIBUTING.md:544]: or -- on a 2nd though -- "whatever!  + we do have `maint` branch used actively , so ok -- let's go with `maint-` prefix". adjusting

File: CONTRIBUTING.md (hunk 9 above)
Comment (yarikoptic) [reply CONTRIBUTING.md]: removed in aca2430ddcde1f68bc34e87ab4c7cbd1bf111b6e -- might resurrect it as a follow up PR for a separate discussion

**Comment by yarikoptic** [comment]
From my PoV it is ready for re-review/merge, please chime in @datalad/developers

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -556,10 +546,37 @@ Refer datalad/config.py for information on how to add these environment variable
...
+- `master`: changes toward the next `MAJOR.MINOR.0` release.
+  Release candidates (tagged with an `rcX` suffix) are cut from this branch
+- `maint`: bug fixes for the latest released `MAJOR.MINOR.PATCH`
+- `maint-MAJOR.MINOR`: generally not used, unless some bug fix release with a critical bug fix is needed.
+
+## Workflow
+
+- upon release of `MAJOR.MINOR.0`, `maint` branch need to be fast-forwarded to that release
+- bug fixes to functionality released within the `maint` branch, should be
+  submitted against `maint ` branch
```
Comment (adswa) [inline CONTRIBUTING.md:562]: I found typos

```suggestion
- upon release of `MAJOR.MINOR.0`, `maint` branch needs to be fast-forwarded to that release
- bug fixes to functionality released within the `maint` branch should be
  submitted against `maint` branch
```
